File: java/src/main/java/neetcode/twopointers/p0125_validpalindrome/Solution.java
package neetcode.twopointers.p0125_validpalindrome;

public class Solution {
    /**
     * Returns true if the string is a palindrome, considering only alphanumeric
     * characters and ignoring case.
     *
     * Approach: two pointers converging from both ends, skipping
     * non-alphanumeric characters in-place. No string allocation needed.
     *
     * Time:  O(n)
     * Space: O(1)
     */
    public boolean isPalindrome(String s) {
        int i = 0;
        int j = s.length() - 1;

        while (i < j) {
            // Skip non-alphanumeric characters from the left
            while (i < j && !Character.isLetterOrDigit(s.charAt(i)))  i++;
            // Skip non-alphanumeric characters from the right
            while (i < j && !Character.isLetterOrDigit(s.charAt(j)))  j--;
            // Compare case-insensitive
            if (Character.toLowerCase(s.charAt(i)) != Character.toLowerCase(s.charAt(j))) {
                return false;
            }
            i++;
            j--;
        }
        return true;
    }
}

File: python/src/neetcode/twopointers/p0125_validpalindrome/solution.py
class Solution:
    """One-liner using Python string tools."""

    def isPalindrome(self, s: str) -> bool:
        """
        Cleans the string (keep only alphanumeric, lowercase),
        then checks if it equals its reverse.

        Time:  O(n)
        Space: O(n) — the cleaned string
        """
        cleaned = "".join(c.lower() for c in s if c.isalnum())
        return cleaned == cleaned[::-1]


class SolutionTwoPointers:
    """Explicit two-pointer approach — O(1) space, mirrors the Java solution."""

    def isPalindrome(self, s: str) -> bool:
        """
        Converging two pointers, skipping non-alphanumeric characters in place.

        Time:  O(n)
        Space: O(1) — no auxiliary string
        """
        left, right = 0, len(s) - 1

        while left < right:
            while left < right and not s[left].isalnum():
                left += 1
            while left < right and not s[right].isalnum():
                right -= 1
            if s[left].lower() != s[right].lower():
                return False
            left += 1
            right -= 1

        return True
